File: app/src/test/java/fmrsabino/moviesdb/DatabaseHelperTest.java
package fmrsabino.moviesdb;

import android.database.Cursor;
import android.os.Build;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import fmrsabino.RxSchedulersOverrideRule;
import fmrsabino.moviesdb.data.local.DatabaseHelper;
import fmrsabino.moviesdb.data.local.DbOpenHelper;
import fmrsabino.moviesdb.data.local.table.ImageTable;
import fmrsabino.moviesdb.data.model.configuration.Image;
import rx.observers.TestSubscriber;

import static junit.framework.Assert.assertEquals;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = Build.VERSION_CODES.M)
public class DatabaseHelperTest {
    private DatabaseHelper databaseHelper;

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        databaseHelper = new DatabaseHelper(new DbOpenHelper(RuntimeEnvironment.application));
    }

    @Test
    public void setImageTest() {
        Image image = Image.builder().baseUrl("mock").build();

        //Verify Image emission
        TestSubscriber<Image> testSubscriber = new TestSubscriber<>();
        databaseHelper.setImage(image).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(image);

        //Verify object storage
        Cursor c = databaseHelper.getDb().query("SELECT * FROM " + ImageTable.TABLE_NAME);
        assertEquals(c.getCount(), 1);
        while (c.moveToNext()) {
            Image img = ImageTable.parseCursor(c);
            assertEquals(image, img);
        }
    }

    @Test
    public void setImageReplacesOldImage() {
        Image oldImage = Image.builder().baseUrl("oldImage").build();
        Image newImage = Image.builder().baseUrl("newImage").build();
        databaseHelper.setImage(oldImage).subscribe(new TestSubscriber<>());
        databaseHelper.setImage(newImage).subscribe(new TestSubscriber<>());
        Cursor c = databaseHelper.getDb().query("SELECT * FROM " + ImageTable.TABLE_NAME);
        assertEquals(c.getCount(), 1);
        while (c.moveToNext()) {
            Image img = ImageTable.parseCursor(c);
            assertEquals(newImage, img);
        }
    }

    @Test
    public void observeImageDoesNotEnd() {
        TestSubscriber<Image> testSubscriber = new TestSubscriber<>();
        databaseHelper.observeImage().subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertNotCompleted();
    }

    @Test
    public void observeImageEmitsValues() {
        Image image = Image.builder().baseUrl("mock").build();
        databaseHelper.setImage(image).subscribe(new TestSubscriber<>());
        TestSubscriber<Image> testSubscriber = new TestSubscriber<>();
        databaseHelper.observeImage().subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(image);
    }
}

File: app/src/test/java/fmrsabino/moviesdb/DatabaseHelperTest.kt
package fmrsabino.moviesdb

import fmrsabino.moviesdb.data.local.DatabaseHelper
import fmrsabino.moviesdb.data.local.DbOpenHelper
import fmrsabino.moviesdb.data.local.table.ImageTable
import fmrsabino.moviesdb.data.model.configuration.Image
import io.reactivex.observers.TestObserver
import junit.framework.Assert.assertEquals
import junit.framework.Assert.assertNotNull
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config


@RunWith(RobolectricTestRunner::class)
@Config(constants = BuildConfig::class)
class DatabaseHelperTest {
    lateinit var databaseHelper: DatabaseHelper

    @JvmField @Rule val trampolineSchedulerRule = TrampolineSchedulerRule()

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        databaseHelper = DatabaseHelper(DbOpenHelper(RuntimeEnvironment.application))
    }

    @Test
    fun setImageTest() {
        val image = Image(baseUrl = "mockUrl")

        //Verify Image emission
        val testObserver = TestObserver<Image>()
        databaseHelper.setImage(image).subscribe(testObserver)
        testObserver.assertNoErrors()
        testObserver.assertValue(image)

        //Verify object storage
        val c = databaseHelper.db.query("SELECT * FROM " + ImageTable.TABLE_NAME)
        assertNotNull(c)
        c?.let {
            assertEquals(it.count, 1)
            it.moveToNext()
            val img = ImageTable.parseCursor(it)
            assertEquals(image, img)
        }
    }

    @Test
    fun setImageReplacesOldImage() {
        val oldImage = Image(baseUrl = "oldImage")
        val newImage = Image(baseUrl = "newImage")
        databaseHelper.setImage(oldImage).subscribe()
        databaseHelper.setImage(newImage).subscribe()
        val c = databaseHelper.db.query("SELECT * FROM " + ImageTable.TABLE_NAME)
        assertNotNull(c)
        c?.let {
            assertEquals(it.count, 1)
            val img = ImageTable.parseCursor(it)
            assertEquals(newImage, img)
        }
    }

    @Test
    fun observeImageDoesNotEnd() {
        val testObserver = TestObserver<Image>()
        databaseHelper.observeImage().subscribe(testObserver)
        testObserver.assertNoErrors()
        testObserver.assertNotTerminated()
    }

    @Test
    fun observeImageEmitsValues() {
        val image = Image()
        databaseHelper.setImage(image).subscribe(TestObserver())
        val testObserver = TestObserver<Image>()
        databaseHelper.observeImage().subscribe(testObserver)
        testObserver.assertNoErrors()
        testObserver.assertValue(image)
        testObserver.assertNotTerminated()
    }
}
